src/treelearner/gpu_tree_learner.cpp

```diff
@@ -0,0 +1,1210 @@
+    }
+    else {
+      kernel_wait_obj_ = boost::compute::wait_list(
+                         queue_.enqueue_1d_range_kernel(histogram_kernels_[exp_workgroups_per_feature], 0, num_workgroups * 256, 256));
+    }
+  }
+  // copy the results asynchronously. Size depends on if double precision is used
+  size_t output_size = num_dense_feature4_ * dword_features_ * device_bin_size_ * hist_bin_entry_sz_;
+  boost::compute::event histogram_wait_event;
+  host_histogram_outputs_ = (void*)queue_.enqueue_map_buffer_async(device_histogram_outputs_, boost::compute::command_queue::map_read, 
+                                                                   0, output_size, histogram_wait_event, kernel_wait_obj_);
+  // we will wait for this object in WaitAndGetHistograms
+  histograms_wait_obj_ = boost::compute::wait_list(histogram_wait_event);
+}
+
+template <typename HistType>
+void GPUTreeLearner::WaitAndGetHistograms(FeatureHistogram* histograms, const std::vector<int8_t>& is_feature_used) {
+  HistType* hist_outputs = (HistType*) host_histogram_outputs_;
+  // when the output is ready, the compuataion is done
+  histograms_wait_obj_.wait();
+  #pragma omp parallel for schedule(static)
+  for(int i = 0; i < num_dense_features_; ++i) {
+    int dense_index = dense_feature_map_[i];
+    if (!is_feature_used[dense_index]) {
+      continue;
+    }
+    auto old_histogram_array = histograms[dense_index].RawData() - 1;
+    int bin_size = train_data_->FeatureNumTotalBin(dense_index); 
+    // printf("Feature %d, Bin = %d, Total Bin = %d\n", dense_index, train_data_->FeatureNumBin(dense_index), train_data_->FeatureNumTotalBin(dense_index));
+    if (device_bin_mults_[i] == 1) {
+      for (int j = 0; j < bin_size; ++j) {
+        old_histogram_array[j].sum_gradients = hist_outputs[i * device_bin_size_+ j].sum_gradients;
+        old_histogram_array[j].sum_hessians = hist_outputs[i * device_bin_size_ + j].sum_hessians;
+        old_histogram_array[j].cnt = hist_outputs[i * device_bin_size_ + j].cnt;
+      }
+    }
+    else {
+      // values of this feature has been redistributed to multiple bins; need a reduction here
+      int ind = 0;
+      for (int j = 0; j < bin_size; ++j) {
+        double sum_g = 0.0, sum_h = 0.0;
+        size_t cnt = 0;
+        for (int k = 0; k < device_bin_mults_[i]; ++k) {
+          sum_g += hist_outputs[i * device_bin_size_+ ind].sum_gradients;
+          sum_h += hist_outputs[i * device_bin_size_+ ind].sum_hessians;
+          cnt += hist_outputs[i * device_bin_size_ + ind].cnt;
+          ind++;
+        }
+        old_histogram_array[j].sum_gradients = sum_g;
+        old_histogram_array[j].sum_hessians = sum_h;
+        old_histogram_array[j].cnt = cnt;
+      }
+    }
+  }
+  queue_.enqueue_unmap_buffer(device_histogram_outputs_, host_histogram_outputs_);
+}
+
+void GPUTreeLearner::AllocateGPUMemory() {
+  num_dense_features_ = 0;
+  for (int i = 0; i < num_features_; ++i) {
```

Comment:
A simple explanation: 
1. When building histogram, you can just use the bin_data in a feature_group to build a histogram. However, this histogram actually is the merging of many features (feature_index are continuous). And the histograms of the features in this group can be restored(By calling `Dataset::FixHistogram`) from this merged histogram. Refer to : https://github.com/Microsoft/LightGBM/blob/master/include/LightGBM/feature_group.h#L29-L49 
2. when finding best split, you should use the original features histograms.